src/script/miniscript.h

```diff
@@ -779,8 +779,11 @@ struct Node {
     //! Check whether there is no satisfaction path that contains both timelocks and heightlocks
     bool CheckTimeLocksMix() const { return GetType() << "k"_mst; }
 
-    //! Do all sanity checks.
-    bool IsSane() const { return IsValid() && IsNonMalleable() && CheckTimeLocksMix() && CheckOpsLimit() && CheckStackSize(); }
+    //! Whether successful non-malleable satisfactions are guaranteed to be valid.
+    bool ValidSatisfactions() const { return IsValid() && CheckOpsLimit() && CheckStackSize(); }
+
+    //! Whether the apparent policy of this node matches its script semantics.
+    bool IsSane() const { return ValidSatisfactions() && IsNonMalleable() && CheckTimeLocksMix(); }
```

Comment:
Technically, all timelocks are malleable if there is no signature. If there is no signature, anyone can malleate nLockTime/nSequence to any value perhaps not matching the script semantics?  

Why is the NeedsSignature check only on top-level? and not a general condition for sanity?

Reply:
You are right, `n:older(12)` shouldn't be called "sane". But to call it "sane" one should use `IsSaneTopLevel()`.

I think that the check for a fragment's sanity should be context-less, and the requirement for signature depends on the context. To take back the example from above `n:older(12)` shouldn't be called sane on its own, but it should in the context of `and_b(n:older(12),s:pk(A))`. So it should only be called insane if not 's' and is the top level fragment. That's what `IsSaneTopLevel()` is.
It's not purely theoretical, for instance we use this in #24148 to find the deepest insane sub of an insane Miniscript: https://github.com/bitcoin/bitcoin/pull/24148/files#diff-a55760aaec4bce216663f5ebf65823516347356a8320d30459427149f7bbc2c5R824-R831. If the `IsSane()` check is context-dependent it gets messy.

Now, maybe there is a point for the need for a better naming.

Reply:
Done.